transport/mixer/src/channel.rs

```diff
@@ -0,0 +1,356 @@
+    type Item = T;
+
+    fn poll_next(self: std::pin::Pin<&mut Self>, cx: &mut std::task::Context<'_>) -> Poll<Option<Self::Item>> {
+        let mut channel = self.channel.lock().unwrap();
+        let now = std::time::SystemTime::now();
+        if channel.sender_count > 0 {
+            if channel.buffer.peek().map(|x| x.0.release_at < now).unwrap_or(false) {
+                let data = channel
+                    .buffer
+                    .pop()
+                    .expect("The value should be present within the same locked access")
+                    .0
+                    .item;
+
+                trace!("yielding a value");
+
+                #[cfg(all(feature = "prometheus", not(test)))]
+                METRIC_QUEUE_SIZE.decrement(1.0f64);
+
+                return Poll::Ready(Some(data));
+            }
+
+            let waker = cx.waker().clone();
+            channel.waker = Some(waker);
+
+            if let Some(next) = channel.buffer.peek() {
+                if let Ok(remaining) = next.0.release_at.duration_since(now) {
+                    trace!("reseting the timer");
+                    channel.timer.reset(remaining);
+                    let timer = std::pin::pin!(&mut channel.timer);
+                    let res = timer.poll(cx);
+
+                    return match res {
+                        Poll::Ready(_) => {
+                            #[cfg(all(feature = "prometheus", not(test)))]
+                            METRIC_QUEUE_SIZE.decrement(1.0f64);
+
+                            Poll::Ready(Some(
+                                channel
+                                    .buffer
+                                    .pop()
+                                    .expect("The value should be present within the locked access")
+                                    .0
+                                    .item,
+                            ))
+                        }
+                        Poll::Pending => Poll::Pending,
+                    };
+                } else {
+                    unreachable!("the previous block would've yielded the value");
+                }
+            }
+
+            Poll::Pending
+        } else {
+            channel.receiver_active = false;
+            Poll::Ready(None)
+        }
+    }
+}
+
+impl<T> Receiver<T> {
+    /// Receive a single delayed mixed item.
+    pub async fn recv(&mut self) -> Option<T> {
+        poll_fn(|cx| self.poll_next_unpin(cx)).await
+    }
+}
+
+/// Instantiate a mixing channel and return the sender and receiver end of the channel.
+pub fn channel<T>() -> (Sender<T>, Receiver<T>) {
+    #[cfg(all(feature = "prometheus", not(test)))]
+    {
+        // Initialize the lazy statics here
+        lazy_static::initialize(&METRIC_QUEUE_SIZE);
+        lazy_static::initialize(&METRIC_MIXER_AVERAGE_DELAY);
+    }
+
+    let mut buffer = BinaryHeap::new();
+    let mixer_capacity = std::env::var("HOPR_INTERNAL_MIXER_CAPACITY")
```

Comment:
I'm very much against library crates parsing any env variables. A better approach is to have a proper config object that gets passed by the caller. And then the upper most binary (= hoprd in our case), gets to decide which are filled from a config file/cli or special environment variables.

Reply:
These should be never set. In the previous versions these have been static variables in the code. I'm not sure these should be set at all. But for testing mechanics, I needed it there.

Reply:
The `MixerConfig` object is not configuring the channel and the default instantiation from the env variables has been lifted to the `protocol` crate, where the mixer channel instantiation occurs until the entire transport construction is liften and unified.